### PR TITLE
System level auth

### PR DESCRIPTION
**Pull request by jj-stigell**
Backend implementation for role based authorization.

Routes that need to be protected with role access rights -> include the authorization middleware. Allowed roles must be defined for the route as an array `Array<UserRole>`.

Following route(s) require role 'ADMIN' in the jwt token payload
- Creating a new course

Any other routes we should protect?

related to #241 and #229

### CODE REVIEW

**Comment by no-comma** [comment]
I'll review this on Wednesday.

File: server/test/controllers/auth.test.ts [hunk 1]
```diff
@@ -60,7 +60,7 @@ describe('Test POST /v1/auth/signup - create a new user', () => {
         name: 'aalto',
         password: 'grades',
         studentID: '123456',
-        role: 'SYSADMIN'
+        role: 'ADMIN'
```
Comment (no-comma) [inline server/test/controllers/auth.test.ts:63]: Use the `enum` instead? Also applies to the other `'ADMIN'` change in this file.

**Review by no-comma — CHANGES_REQUESTED** [review]
I had a chance to look at this now already, and other than one minor detail everything looks good.

File: client/src/types/general.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// SPDX-FileCopyrightText: 2023 The Aalto Grades Developers
+//
+// SPDX-License-Identifier: MIT
+
+/* eslint-disable no-unused-vars */
+export enum UserRole {
+  User = 'USER',
+  Admin = 'ADMIN'
+}
```
Comment (no-comma) [inline client/src/types/general.ts]: This works for this PR, but I would like to have a single code module for types shared by the frontend and backend, so we don't need to manually keep them in sync or duplicate code between `server` and `client`.

Comment (jj-stigell) [reply client/src/types/general.ts]: So something like folder "common" to the project root and subfolder "types" where we place client/server shared types?

Comment (jj-stigell) [reply client/src/types/general.ts]: So somehow with declaring global types and no need for importing. I will leave this for now as is, let's think this one later.

File: client/src/components/FrontPage.tsx [hunk 3]
```diff
@@ -31,9 +35,16 @@ const FrontPage = (): JSX.Element => {
...
           Your Current Courses
         </Typography>
+        { /* admins are shown the button for creating a new course */
+          auth.role == SystemRole.Admin &&
+          <Button id='ag_new_course_btn' size='large' variant='contained' onClick={() => {
+            navigate('/create-course');
+          }}>
+            Create New Course
+          </Button>
+        }
```
Comment (no-comma) [inline client/src/components/FrontPage.tsx:45]: Courses are created from `AllCoursesView` as of #254, so this should be removed.

File: client/src/tests/FrontPage.test.tsx [hunk 4]
```diff
@@ -51,4 +52,20 @@ describe('Tests for FrontPage component', () => {
...
+
+
+  test('FrontPage should render Create New Course for admin', async () => {
+
+    const auth = { role: SystemRole.Admin };
+    renderFrontPage(auth);
+    await waitFor(() => expect(screen.queryByText('Create New Course')).toBeInTheDocument());
+
+  });
+
```
Comment (no-comma) [inline client/src/tests/FrontPage.test.tsx:70]: Same here.

**Review by no-comma — CHANGES_REQUESTED** [review]

**Review by no-comma — APPROVED** [review]
LGTM

**Pull request merged**